File: app/src/main/java/id/ac/stiki/doleno/absenin/di/FirestoreModule.java
package id.ac.stiki.doleno.absenin.di;

import com.google.firebase.firestore.FirebaseFirestore;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.components.ViewModelComponent;
import id.ac.stiki.doleno.absenin.data.source.network.AbsentStore;
import id.ac.stiki.doleno.absenin.data.source.network.EventStore;
import id.ac.stiki.doleno.absenin.data.source.network.UserStore;
import id.ac.stiki.doleno.absenin.data.source.network.impl.AbsentStoreImpl;
import id.ac.stiki.doleno.absenin.data.source.network.impl.EventStoreImpl;
import id.ac.stiki.doleno.absenin.data.source.network.impl.UserStoreImpl;

@Module
@InstallIn(ViewModelComponent.class)
public class FirestoreModule {
    @Provides
    public static FirebaseFirestore provideFireStore() {
        return FirebaseFirestore.getInstance();
    }

    @Provides
    public UserStore provideUserStore(FirebaseFirestore firestore) {
        return new UserStoreImpl(firestore);
    }

    @Provides
    public EventStore provideEventStore(FirebaseFirestore firestore) {
        return new EventStoreImpl(firestore) {
        };
    }

    @Provides
    public AbsentStore provideAbsentStore(FirebaseFirestore firestore) {
        return new AbsentStoreImpl(firestore);
    }
}

File: app/src/main/java/id/ac/stiki/doleno/absenin/di/FirestoreModule.kt
package id.ac.stiki.doleno.absenin.di

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreSettings
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ViewModelComponent
import id.ac.stiki.doleno.absenin.data.source.network.*
import id.ac.stiki.doleno.absenin.data.source.network.impl.*

@Module
@InstallIn(ViewModelComponent::class)
class FirestoreModule {
    @Provides
    fun provideUserStore(firestore: FirebaseFirestore): UserStore {
        return UserStoreImpl(firestore)
    }

    @Provides
    fun provideEventStore(firestore: FirebaseFirestore): EventStore {
        return object : EventStoreImpl(firestore) {}
    }

    @Provides
    fun provideAbsentStore(firestore: FirebaseFirestore): AbsentStore {
        return AbsentStoreImpl(firestore)
    }

    @Provides
    fun provideEventParticipantStore(firestore: FirebaseFirestore): EventParticipantStore {
        return EventParticipantStoreImpl(firestore)
    }

    @Provides
    fun provideGlobalStore(firestore: FirebaseFirestore): GlobalStore {
        return GlobalStoreImpl(firestore)
    }

    companion object {
        @JvmStatic
        @Provides
        fun provideFireStore(): FirebaseFirestore {
            val settings = FirebaseFirestoreSettings.Builder()
            settings.isPersistenceEnabled = false
            val firestore = FirebaseFirestore.getInstance()
            firestore.firestoreSettings = settings.build()
            return firestore
        }
    }
}